<a href="https://colab.research.google.com/github/Chandradithya8/Advanced_Regression/blob/main/Advanced_regression2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [48]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.linear_model import Lasso
from sklearn.ensemble import RandomForestRegressor
from xgboost import XGBRegressor
from sklearn.feature_selection import SelectFromModel
from sklearn.metrics import r2_score

Feature engineering for train dataset

In [2]:
df_train=pd.read_csv("/content/train_ad.csv")
df_train.head()

Id  MSSubClass MSZoning  ...  SaleType  SaleCondition SalePrice
0   1          60       RL  ...        WD         Normal    208500
1   2          20       RL  ...        WD         Normal    181500
2   3          60       RL  ...        WD         Normal    223500
3   4          70       RL  ...        WD        Abnorml    140000
4   5          60       RL  ...        WD         Normal    250000

[5 rows x 81 columns]

In [3]:
## splitting the categorical variables
categories=[features for features in df_train.columns if df_train[features].dtypes=='O']
print(len(categories))
print(categories)

43
['MSZoning', 'Street', 'Alley', 'LotShape', 'LandContour', 'Utilities', 'LotConfig', 'LandSlope', 'Neighborhood', 'Condition1', 'Condition2', 'BldgType', 'HouseStyle', 'RoofStyle', 'RoofMatl', 'Exterior1st', 'Exterior2nd', 'MasVnrType', 'ExterQual', 'ExterCond', 'Foundation', 'BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2', 'Heating', 'HeatingQC', 'CentralAir', 'Electrical', 'KitchenQual', 'Functional', 'FireplaceQu', 'GarageType', 'GarageFinish', 'GarageQual', 'GarageCond', 'PavedDrive', 'PoolQC', 'Fence', 'MiscFeature', 'SaleType', 'SaleCondition']


In [4]:
## categorical variables with nan values
categories_na=[features for features in categories if  df_train[features].isnull().sum()>0]
len(categories_na)

16

In [5]:
## replacing nan values with new features
for features in categories_na:
  df_train[features]=np.where(df_train[features].isnull(),'Missing',df_train[features])
df_train[categories_na].isnull().sum()  

Alley           0
MasVnrType      0
BsmtQual        0
BsmtCond        0
BsmtExposure    0
BsmtFinType1    0
BsmtFinType2    0
Electrical      0
FireplaceQu     0
GarageType      0
GarageFinish    0
GarageQual      0
GarageCond      0
PoolQC          0
Fence           0
MiscFeature     0
dtype: int64

In [6]:
## splitting the numerical variables
numerical=[features for features in df_train.columns if features not in categories]
len(numerical)

38

In [7]:
## splitting the numerical variables with nan values
numerical_na=[features for features in numerical if df_train[features].isnull().sum()>0]
len(numerical_na)

3

In [8]:
## replacing nan values with the median
for features in numerical_na:
  df_train[features]=np.where(df_train[features].isnull(),df_train[features].median(),df_train[features])
df_train[numerical_na].isnull().sum()  

LotFrontage    0
MasVnrArea     0
GarageYrBlt    0
dtype: int64

In [9]:
df_train.isnull().sum()

Id               0
MSSubClass       0
MSZoning         0
LotFrontage      0
LotArea          0
                ..
MoSold           0
YrSold           0
SaleType         0
SaleCondition    0
SalePrice        0
Length: 81, dtype: int64

In [10]:
## splitting the temporal variables(date time variables)
temporal=[features for features in numerical if 'Yr' in features or 'Year' in features]
temporal

['YearBuilt', 'YearRemodAdd', 'GarageYrBlt', 'YrSold']

In [11]:
for features in temporal:
  if 'YrSold' not in features:
    df_train[features]=df_train['YrSold']-df_train[features]
df_train[numerical].head()    

Id  MSSubClass  LotFrontage  LotArea  ...  MiscVal  MoSold  YrSold  SalePrice
0   1          60         65.0     8450  ...        0       2    2008     208500
1   2          20         80.0     9600  ...        0       5    2007     181500
2   3          60         68.0    11250  ...        0       9    2008     223500
3   4          70         60.0     9550  ...        0       2    2006     140000
4   5          60         84.0    14260  ...        0      12    2008     250000

[5 rows x 38 columns]

In [12]:
for features in ['LotFrontage', 'LotArea', '1stFlrSF', 'GrLivArea', 'SalePrice']:
  if 0 in df_train[features].unique():
    pass
  else:
    df_train[features]=np.log(df_train[features])
df_train[numerical].head()

Id  MSSubClass  LotFrontage   LotArea  ...  MiscVal  MoSold  YrSold  SalePrice
0   1          60     4.174387  9.041922  ...        0       2    2008  12.247694
1   2          20     4.382027  9.169518  ...        0       5    2007  12.109011
2   3          60     4.219508  9.328123  ...        0       9    2008  12.317167
3   4          70     4.094345  9.164296  ...        0       2    2006  11.849398
4   5          60     4.430817  9.565214  ...        0      12    2008  12.429216

[5 rows x 38 columns]

In [13]:
## handling the categorical values
for features in categories:
  data_features=df_train.groupby(features)['SalePrice'].mean().sort_values().index
  labels_ordered={i:k for k,i in enumerate(data_features,0)}
  df_train[features]=df_train[features].map(labels_ordered)

In [14]:
df_train.head()

Id  MSSubClass  MSZoning  ...  SaleType  SaleCondition  SalePrice
0   1          60         3  ...         4              4  12.247694
1   2          20         3  ...         4              4  12.109011
2   3          60         3  ...         4              4  12.317167
3   4          70         3  ...         4              1  11.849398
4   5          60         3  ...         4              4  12.429216

[5 rows x 81 columns]

In [15]:
x_train=df_train.drop(['SalePrice','Id'],axis=1)
y_train=df_train['SalePrice']

In [16]:
## feature scaling-StandardScaler
scaler=StandardScaler()
scaling=scaler.fit_transform(x_train)
scaling

array([[ 0.07337496,  0.35904396, -0.07586857, ...,  0.13877749,
        -0.22029532,  0.18996938],
       [-0.87256276,  0.35904396,  0.57242366, ..., -0.61443862,
        -0.22029532,  0.18996938],
       [ 0.07337496,  0.35904396,  0.06500658, ...,  0.13877749,
        -0.22029532,  0.18996938],
       ...,
       [ 0.30985939,  0.35904396, -0.02820043, ...,  1.64520971,
        -0.22029532,  0.18996938],
       [-0.87256276,  0.35904396,  0.06500658, ...,  1.64520971,
        -0.22029532,  0.18996938],
       [-0.87256276,  0.35904396,  0.3709213 , ...,  0.13877749,
        -0.22029532,  0.18996938]])

In [17]:
x_train=pd.DataFrame(scaling,columns=x_train.columns)
x_train.head()

MSSubClass  MSZoning  LotFrontage  ...    YrSold  SaleType  SaleCondition
0    0.073375  0.359044    -0.075869  ...  0.138777 -0.220295       0.189969
1   -0.872563  0.359044     0.572424  ... -0.614439 -0.220295       0.189969
2    0.073375  0.359044     0.065007  ...  0.138777 -0.220295       0.189969
3    0.309859  0.359044    -0.325778  ... -1.367655 -0.220295      -3.192412
4    0.073375  0.359044     0.724756  ...  0.138777 -0.220295       0.189969

[5 rows x 79 columns]

In [18]:
## handling the outliers
for features in x_train.columns:
  mean=x_train[features].mean()
  std=x_train[features].std()
  upper_bound=mean+(3*std)
  lower_bound=mean-(3*std)
  x_train[features]=np.where(x_train[features]>upper_bound,upper_bound,x_train[features])
  x_train[features]=np.where(x_train[features]<lower_bound,lower_bound,x_train[features])

In [19]:
x_train.head()

MSSubClass  MSZoning  LotFrontage  ...    YrSold  SaleType  SaleCondition
0    0.073375  0.359044    -0.075869  ...  0.138777 -0.220295       0.189969
1   -0.872563  0.359044     0.572424  ... -0.614439 -0.220295       0.189969
2    0.073375  0.359044     0.065007  ...  0.138777 -0.220295       0.189969
3    0.309859  0.359044    -0.325778  ... -1.367655 -0.220295      -3.001028
4    0.073375  0.359044     0.724756  ...  0.138777 -0.220295       0.189969

[5 rows x 79 columns]

Feature Engineering for test dataset

In [20]:
df_test=pd.read_csv("/content/train_ad.csv")
df_test.head()

Id  MSSubClass MSZoning  ...  SaleType  SaleCondition SalePrice
0   1          60       RL  ...        WD         Normal    208500
1   2          20       RL  ...        WD         Normal    181500
2   3          60       RL  ...        WD         Normal    223500
3   4          70       RL  ...        WD        Abnorml    140000
4   5          60       RL  ...        WD         Normal    250000

[5 rows x 81 columns]

In [21]:
## splitting the categorical data
categories=[features for features in df_test.columns if df_test[features].dtypes=='O']
len(categories)

43

In [22]:
## splitting the categorical data having nan values
categories_na=[features for features in categories if df_test[features].isnull().sum()>0]
len(categories_na)

16

In [23]:
## replacing nan values with new values
for features in categories_na:
  df_test[features]=np.where(df_test[features].isnull(),'Missing',df_test[features])
df_test[categories_na].isnull().sum()  

Alley           0
MasVnrType      0
BsmtQual        0
BsmtCond        0
BsmtExposure    0
BsmtFinType1    0
BsmtFinType2    0
Electrical      0
FireplaceQu     0
GarageType      0
GarageFinish    0
GarageQual      0
GarageCond      0
PoolQC          0
Fence           0
MiscFeature     0
dtype: int64

In [24]:
## splitting the data with numerical values
numerical=[features for features in df_test.columns if features not in categories]
len(numerical)

38

In [25]:
## splitting the  numerical data with nan values
numerical_na=[features for features in numerical if df_test[features].isnull().sum()>0]
len(numerical_na)

3

In [26]:
## replacing the nan values with median value
for features in numerical_na:
  df_test[features]=np.where(df_test[features].isnull(),df_test[features].median(),df_test[features])
df_test[numerical_na].isnull().sum()  

LotFrontage    0
MasVnrArea     0
GarageYrBlt    0
dtype: int64

In [27]:
## splitting the temporal data
temporal=[features for features in numerical if 'Yr' in features or 'Year' in features]
temporal

['YearBuilt', 'YearRemodAdd', 'GarageYrBlt', 'YrSold']

In [28]:
for features in temporal:
  if 'YrSold' not in features:
    df_test[features]=df_test['YrSold']-df_test[features]
df_test.head()   

Id  MSSubClass MSZoning  ...  SaleType  SaleCondition SalePrice
0   1          60       RL  ...        WD         Normal    208500
1   2          20       RL  ...        WD         Normal    181500
2   3          60       RL  ...        WD         Normal    223500
3   4          70       RL  ...        WD        Abnorml    140000
4   5          60       RL  ...        WD         Normal    250000

[5 rows x 81 columns]

In [29]:
for features in ['LotFrontage', 'LotArea', '1stFlrSF', 'GrLivArea', 'SalePrice']:
  if 0 in df_test[features].unique():
    pass
  else:
    df_test[features]=np.log(df_test[features])
df_test[numerical].head()    

Id  MSSubClass  LotFrontage   LotArea  ...  MiscVal  MoSold  YrSold  SalePrice
0   1          60     4.174387  9.041922  ...        0       2    2008  12.247694
1   2          20     4.382027  9.169518  ...        0       5    2007  12.109011
2   3          60     4.219508  9.328123  ...        0       9    2008  12.317167
3   4          70     4.094345  9.164296  ...        0       2    2006  11.849398
4   5          60     4.430817  9.565214  ...        0      12    2008  12.429216

[5 rows x 38 columns]

In [30]:
## handling the categorical values
for features in categories:
  data_features=df_test.groupby(features)['SalePrice'].mean().sort_values().index
  dictionary={k:i for i,k in enumerate(data_features,0)}
  df_test[features]=df_test[features].map(dictionary)

In [31]:
df_test.head()

Id  MSSubClass  MSZoning  ...  SaleType  SaleCondition  SalePrice
0   1          60         3  ...         4              4  12.247694
1   2          20         3  ...         4              4  12.109011
2   3          60         3  ...         4              4  12.317167
3   4          70         3  ...         4              1  11.849398
4   5          60         3  ...         4              4  12.429216

[5 rows x 81 columns]

In [32]:
x=df_test.drop(['SalePrice','Id'],axis=1)
y=df_test['SalePrice']

In [33]:
## feature scaling
scaler=StandardScaler()
scaling=scaler.fit_transform(x)
scaling

array([[ 0.07337496,  0.35904396, -0.07586857, ...,  0.13877749,
        -0.22029532,  0.18996938],
       [-0.87256276,  0.35904396,  0.57242366, ..., -0.61443862,
        -0.22029532,  0.18996938],
       [ 0.07337496,  0.35904396,  0.06500658, ...,  0.13877749,
        -0.22029532,  0.18996938],
       ...,
       [ 0.30985939,  0.35904396, -0.02820043, ...,  1.64520971,
        -0.22029532,  0.18996938],
       [-0.87256276,  0.35904396,  0.06500658, ...,  1.64520971,
        -0.22029532,  0.18996938],
       [-0.87256276,  0.35904396,  0.3709213 , ...,  0.13877749,
        -0.22029532,  0.18996938]])

In [34]:
x=pd.DataFrame(scaling,columns=x.columns)
x.head()

MSSubClass  MSZoning  LotFrontage  ...    YrSold  SaleType  SaleCondition
0    0.073375  0.359044    -0.075869  ...  0.138777 -0.220295       0.189969
1   -0.872563  0.359044     0.572424  ... -0.614439 -0.220295       0.189969
2    0.073375  0.359044     0.065007  ...  0.138777 -0.220295       0.189969
3    0.309859  0.359044    -0.325778  ... -1.367655 -0.220295      -3.192412
4    0.073375  0.359044     0.724756  ...  0.138777 -0.220295       0.189969

[5 rows x 79 columns]

In [35]:
## handling the outliers
for features in x.columns:
  mean=x[features].mean()
  std=x[features].std()
  upper_bound=mean+(3*std)
  lower_bound=mean-(3*std)
  x[features]=np.where(x[features]>upper_bound,upper_bound,x[features])
  x[features]=np.where(x[features]<lower_bound,lower_bound,x[features])

In [36]:
x.head()

MSSubClass  MSZoning  LotFrontage  ...    YrSold  SaleType  SaleCondition
0    0.073375  0.359044    -0.075869  ...  0.138777 -0.220295       0.189969
1   -0.872563  0.359044     0.572424  ... -0.614439 -0.220295       0.189969
2    0.073375  0.359044     0.065007  ...  0.138777 -0.220295       0.189969
3    0.309859  0.359044    -0.325778  ... -1.367655 -0.220295      -3.001028
4    0.073375  0.359044     0.724756  ...  0.138777 -0.220295       0.189969

[5 rows x 79 columns]

In [37]:
x.shape

(1460, 79)

In [38]:
## the greater the alpha value the less number of features will be selected
selected=SelectFromModel(Lasso(alpha=0.005,random_state=0))
selected.fit(x_train,y_train)

SelectFromModel(estimator=Lasso(alpha=0.005, copy_X=True, fit_intercept=True,
                                max_iter=1000, normalize=False, positive=False,
                                precompute=False, random_state=0,
                                selection='cyclic', tol=0.0001,
                                warm_start=False),
                max_features=None, norm_order=1, prefit=False, threshold=None)

In [39]:
selected.get_support()

array([ True,  True, False,  True, False, False, False,  True, False,
        True, False,  True,  True, False, False, False,  True,  True,
        True,  True, False, False,  True, False, False, False,  True,
        True,  True,  True, False,  True, False,  True, False, False,
       False,  True, False,  True,  True, False,  True, False, False,
        True,  True, False,  True,  True, False,  True,  True, False,
        True,  True,  True, False, False,  True,  True,  True,  True,
       False,  True,  True, False, False, False,  True, False, False,
       False, False, False, False, False,  True,  True])

In [42]:
selected_features=x_train.columns[selected.get_support()]
selected_features

Index(['MSSubClass', 'MSZoning', 'LotArea', 'LandContour', 'LotConfig',
       'Neighborhood', 'Condition1', 'OverallQual', 'OverallCond', 'YearBuilt',
       'YearRemodAdd', 'Exterior1st', 'ExterQual', 'ExterCond', 'Foundation',
       'BsmtQual', 'BsmtExposure', 'BsmtFinSF1', 'TotalBsmtSF', 'HeatingQC',
       'CentralAir', '1stFlrSF', 'GrLivArea', 'BsmtFullBath', 'FullBath',
       'HalfBath', 'KitchenAbvGr', 'KitchenQual', 'Functional', 'Fireplaces',
       'FireplaceQu', 'GarageFinish', 'GarageCars', 'GarageArea', 'GarageQual',
       'PavedDrive', 'WoodDeckSF', 'ScreenPorch', 'SaleType', 'SaleCondition'],
      dtype='object')

In [45]:
x_train=x_train[selected_features]
x=x[selected_features]
x_train.shape

(1460, 40)

In [41]:
model1=LinearRegression()
model2=RandomForestRegressor()
model3=XGBRegressor()

In [47]:
train1=model1.fit(x_train,y_train)
train1

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [49]:
y_pred=train1.predict(x)
print(r2_score(y,y_pred))

0.9131817966777542


In [54]:
train2=model2.fit(x_train,y_train)
train2

RandomForestRegressor(bootstrap=True, ccp_alpha=0.0, criterion='mse',
                      max_depth=None, max_features='auto', max_leaf_nodes=None,
                      max_samples=None, min_impurity_decrease=0.0,
                      min_impurity_split=None, min_samples_leaf=1,
                      min_samples_split=2, min_weight_fraction_leaf=0.0,
                      n_estimators=100, n_jobs=None, oob_score=False,
                      random_state=None, verbose=0, warm_start=False)

In [55]:
y_pred=train2.predict(x)
print(r2_score(y,y_pred))

0.9844191179683168


In [56]:
train3=model3.fit(x_train,y_train)
train3

[05:50:54] WARNING: /workspace/src/objective/regression_obj.cu:152: reg:linear is now deprecated in favor of reg:squarederror.


XGBRegressor(base_score=0.5, booster='gbtree', colsample_bylevel=1,
             colsample_bynode=1, colsample_bytree=1, gamma=0,
             importance_type='gain', learning_rate=0.1, max_delta_step=0,
             max_depth=3, min_child_weight=1, missing=None, n_estimators=100,
             n_jobs=1, nthread=None, objective='reg:linear', random_state=0,
             reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=None,
             silent=None, subsample=1, verbosity=1)

In [57]:
y_pred=train3.predict(x)
print(r2_score(y,y_pred))

0.9516908220314269


From the above models RandomForestRegressor gives the highest accuracy of 98% . LinearRegression gives a accuracy of 91% and xgboostRegressor gives a accuracy of 95%. 